### PR TITLE
feat: add webUtils module with getPathForFile method

### PR DESCRIPTION
**Pull request by MarshallOfSound**
This is designed to replace the File.path augmentation we currently have in place to allow apps to get the filesystem path for a file that blink has a representation of.

File.path is non-standard and messes with certain websites, using a method like this is effectively 0-cost and removes one of the final deviations we have with web standards.

TODO:
* Bikeshed of `blinkUtils` namespace
* Bikeshed over `getPathForFile` method name
* Tests

Notes: Added new `webUtils.getPathForFile` method to replace `File.path` augmentation

### CODE REVIEW

File: docs/api/blink-utils.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+## Methods
+
+The `blinkUtils` module has the following methods:
+
+### `blinkUtils.getPathForFile(file)`
+
+* `file` File - A web [File](https://developer.mozilla.org/en-US/docs/Web/API/File) object.
+
+Returns `string` - The file system path that this `File` object points to. In the case where the object passed in is not a `File` object an exception is thrown. In the case where the File object passed in was constructed in JS and is not backed by a file on disk an empty string is returned.
```
Comment (ckerr) [inline docs/api/blink-utils.md:15]: Is there anything we need to be doing to throw an exception when the object is not a `File`?

**Review by ckerr — COMMENTED** [review]
What's here so far LGTM.

Agree on the TODO for adding tests. I have no opinion on the naming but surely the API WG will want to bikeshed :smile_cat:

File: shell/renderer/api/electron_api_blink_utils.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    return "";
+  }
+  return blob.Path();
+}
+
+}  // namespace blink_utils
+
+}  // namespace api
+
+}  // namespace electron
```
Comment (dsanders11) [inline shell/renderer/api/electron_api_blink_utils.cc:33]: Nest that namespace bad boy.

```suggestion
namespace electron::api::blink_utils {

std::string GetPathForFile(gin_helper::Arguments* args,
                           v8::Local<v8::Value> file) {
  blink::WebBlob blob = blink::WebBlob::FromV8Value(file);
  if (blob.IsNull()) {
    args->ThrowError(
        "getPathForFile expected to receive a File object but one was not "
        "provided");
    return "";
  }
  return blob.Path();
}

}  // namespace electron::api::blink_utils
```

**Review by dsanders11 — COMMENTED** [review]
Another todo: note deprecation of `.path` in `docs/breaking-changes.md`

File: shell/renderer/api/electron_api_blink_utils.cc (hunk 2 above)
Comment (MarshallOfSound) [reply shell/renderer/api/electron_api_blink_utils.cc:33]: This is the pattern we use elsewhere, not sure why, I assume it's in the chromium styleguide 🤷

Comment (dsanders11) [reply shell/renderer/api/electron_api_blink_utils.cc:33]: 🤔 You used the nested namespaces in `shell/renderer/api/electron_api_blink_utils.h` though?

We should be using nested namespaces everywhere after #37855.

Comment (MarshallOfSound) [reply shell/renderer/api/electron_api_blink_utils.cc:33]: Weird, idk how, I copied these from template files 🤔 Will clean it up

Comment (ckerr) [reply shell/renderer/api/electron_api_blink_utils.cc:33]: I don't see anything in the style guide, but `rg -t h namespace | rg ::` indicates that  upstream is fine with the terser form of declaring nested namespaces :smile_cat:, e.g.

```sh
extensions/browser/browsertest_util.h:namespace extensions::browsertest_util {
extensions/browser/browsertest_util.h:}  // namespace extensions::browsertest_util
extensions/browser/extension_function_histogram_value.h:namespace extensions::functions {
extensions/browser/extension_function_histogram_value.h:}  // namespace extensions::functions
extensions/browser/api/extensions_api_client.h:      std::map<settings_namespace::Namespace, ValueStoreCache*>* caches);
extensions/browser/api/app_runtime/app_runtime_api.h:namespace api::app_runtime {
extensions/browser/api/app_runtime/app_runtime_api.h:}  // namespace api::app_runtime
```

Comment (dsanders11) [reply shell/renderer/api/electron_api_blink_utils.cc:33]: @ckerr, it's [squirreled away in the "Modern C++ Usage in Chromium" part of their style guide](https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/c++-features.md#nested-namespaces-allowed), with a link to their discussion thread, where the TL;DR is they approved it for usage but didn't feel the need to go do a mass cleanup of existing code, just use it going forward on new code.

**Comment by dsanders11** [comment]
> Bikeshed of `blinkUtils` namespace

I'm not in API WG but my 2 cents on the namespace would be not including blink since that feels a bit too internal/unnecessary detail for end API users.

Possible alternatives:
* `webUtils`
* `rendererUtils`

File: shell/renderer/api/electron_api_blink_utils.cc [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace api {
+
+namespace blink_utils {
+
+std::string GetPathForFile(gin_helper::Arguments* args,
+                           v8::Local<v8::Value> file) {
+  blink::WebBlob blob = blink::WebBlob::FromV8Value(file);
+  if (blob.IsNull()) {
+    args->ThrowError(
```
Comment (zcbenz) [inline shell/renderer/api/electron_api_blink_utils.cc:21]: Better throw a TypeError with `gin_helper::ErrorThrower`.

**Review by zcbenz — APPROVED** [review]
API LGTM

The API name is fine to me, but I'm also good with other alternatives.

**Comment by jkleinsc** [comment]
> > Bikeshed of `blinkUtils` namespace
> 
> I'm not in API WG but my 2 cents on the namespace would be not including blink since that feels a bit too internal/unnecessary detail for end API users.
> 
> Possible alternatives:
> 
> * `webUtils`
> * `rendererUtils`

I agree that `blinkUtils` is a bit esoteric.  I think `rendererUtils` makes sense.

File: docs/api/file-object.md [hunk 4]
```diff
@@ -2,6 +2,11 @@
 
 > Use the HTML5 `File` API to work natively with files on the filesystem.
 
+> **Warning**
+> The `path` property that Electron adds to the `File` interface is deprecated
+> and **will** be removed in a future Electron release.  We reccomend you
```
Comment (itsananderson) [inline docs/api/file-object.md:7]: sp: "recommend"

**Review by itsananderson — COMMENTED** [review]
The old `File.path` augmentation always struck me as a weird privacy gap because AFAIK there's no way to prevent an "untrusted" renderer from accessing the full path of an uploaded file. If we're going to deprecate the old API, would it be worth considering the privacy implications of the new API, and perhaps having a way to limit its availability? Having a way to enable it in "preload" but disable it in the main world might be useful, for example.

**Comment by MarshallOfSound** [comment]
> If we're going to deprecate the old API, would it be worth considering the privacy implications of the new API, and perhaps having a way to limit its availability? Having a way to enable it in "preload" but disable it in the main world might be useful, for example.

At it's core the file path is available to the renderer process (this just returns a value that blink already had).  Following our standard security assumptions this doesn't change the availability of that information to the process.  All it does is improve web compatibility with things that don't expect `File.path` to be present and prevent zero-exploit exposure of that path.

Any permission model we put in place would be self-governing because it would be the renderer guarding itself (which doesn't make sense)

**Comment by miniak** [comment]
@MarshallOfSound does it make sense to expose this in sandboxed renderers as well?

**Comment by MarshallOfSound** [comment]
> @MarshallOfSound does it make sense to expose this in sandboxed renderers as well?

Per my comment above

> At it's core the file path is available to the renderer process (this just returns a value that blink already had).

This isn't exposing any new capabilities to the process, so I think exposing this information to the slightly-more-privileged isolated preload context is perfectly valid 👍

**Comment by miniak** [comment]
@MarshallOfSound what I mean is that the PR is not doing that, it's not listed in lib/sandboxed_renderer/api/module-list.ts. Even though the path won't be directly usable there due to the sandbox, it might still be useful?

**Comment by miniak** [comment]
@MarshallOfSound in case it's added there, it should be listed in docs/tutorial/sandbox.md as added by #38568

**Comment by MarshallOfSound** [comment]
> Even though the path won't be directly usable there due to the sandbox, it might still be useful?

Yeah I think it will still be useful because you could take that path (albeit untrusted) and hand it off to the main process to do something with 🤔

**Comment by MarshallOfSound** [comment]
Thanks for the reviews folks @dsanders11 @ckerr @itsananderson @zcbenz @miniak 

I've renamed the module to the suggested `webUtils` as it's a good home for interactions with "Web" APIs (File, Blob, Etc.).  I think I've addressed all the other comments too 👍

**Review by miniak — APPROVED** [review]

**Comment by miniak** [comment]
API LGTM

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Requesting changes so that this doesn't get merged without tests.

File: shell/common/node_bindings.cc [hunk 5]
```diff
@@ -88,6 +88,7 @@
   V(electron_common_v8_util)
 
 #define ELECTRON_RENDERER_BINDINGS(V) \
+  V(electron_renderer_web_utils)      \
```
Comment (dsanders11) [inline shell/common/node_bindings.cc:91]: nit: alpha sort this list?

**Review by dsanders11 — APPROVED** [review]
nit: `#include <string>` should be added to `shell/renderer/api/electron_api_web_utils.{cc,h}`

**Comment by MarshallOfSound** [comment]
Both test failures are known flakes, merging

**Pull request merged**